{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 240831: Make pe.py script\n",
    "- Obtain PE saturation mutagenesis libraries from PrimeDesign\n",
    "- Generate epegRNA motif linkers using pegLIT.py\n",
    "- Distill PE library into spacer and PBS motifs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PrimeDesign: Command line interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "240831.ipynb\n",
      "pe.py\n",
      "pegLIT.py\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('ls')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[12, 13, 14, 15, 16]\n",
      "12 13 14 15 16\n"
     ]
    }
   ],
   "source": [
    "ls = [12,13,14,15,16]\n",
    "print(str(ls))\n",
    "ar = np.array(ls)\n",
    "print(str(ar)[1:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' PrimeDesign: Run PrimeDesign using Docker\n",
    "        file (str): Input file (.txt or .csv) with sequences for PrimeDesign. Format: target_name,target_sequence (column names required)\n",
    "        pbs_length_list (list): List of primer binding site (PBS) lengths for the pegRNA extension. Example: 12 13 14 15\n",
    "        rtt_length_list (list): List of reverse transcription (RT) template lengths for the pegRNA extension. Example: 10 15 20\n",
    "        nicking_distance_minimum (int): Minimum nicking distance for designing ngRNAs. (Default: 0 bp)\n",
    "        nicking_distance_maximum (int): Maximum nicking distance for designing ngRNAs. (Default: 100 bp)\n",
    "        filter_c1_extension (bool): Filter against pegRNA extensions that start with a C base. (Default: False)\n",
    "        silent_mutation (bool): Introduce silent mutation into PAM assuming sequence is in-frame. Currently only available with SpCas9. (Default: False)\n",
    "        genome_wide_design (bool): Whether this is a genome-wide pooled design. This option designs a set of pegRNAs per input without ranging PBS and RTT parameters.\n",
    "        saturation_mutagenesis (str): Saturation mutagenesis design with prime editing (Options: 'aa', 'base').\n",
    "        number_of_pegrnas (int): Maximum number of pegRNAs to design for each input sequence. The pegRNAs are ranked by 1) PAM disrupted > PAM intact then 2) distance to edit. (Default: 3)\n",
    "        number_of_ngrnas (int): Maximum number of ngRNAs to design for each input sequence. The ngRNAs are ranked by 1) PE3b-seed > PE3b-nonseed > PE3 then 2) deviation from nicking_distance_pooled. (Default: 3)\n",
    "        nicking_distance_pooled (int): The nicking distance between pegRNAs and ngRNAs for pooled designs. PE3b annotation is priority (PE3b seed -> PE3b non-seed), followed by nicking distance closest to this parameter. (Default: 75 bp)\n",
    "        homology_downstream (int): For pooled designs (genome_wide or saturation_mutagenesis needs to be indicated), this parameter determines the RT extension length downstream of an edit for pegRNA designs. (Default: 10)\n",
    "        pbs_length_pooled (int): The PBS length to design pegRNAs for pooled design applications. (Default: 14 nt)\n",
    "        rtt_max_length_pooled (int): Maximum RTT length to design pegRNAs for pooled design applications. (Default: 50 nt)\n",
    "        out_dir (str): Name of output directory. (Default: ./DATETIMESTAMP_PrimeDesign)\n",
    "    Dependencies: os,numpy,https://github.com/pinellolab/PrimeDesign\n",
    "'''\n",
    "def PrimeDesign(file: str,pbs_length_list: list = [],rtt_length_list: list = [],nicking_distance_minimum: int = 0,\n",
    "                nicking_distance_maximum: int = 100,filter_c1_extension: bool = False,silent_mutation: bool = False,\n",
    "                genome_wide_design: bool = False,saturation_mutagenesis: str = None,number_of_pegrnas: int = 3,number_of_ngrnas: int = 3,\n",
    "                nicking_distance_pooled: int = 75,homology_downstream: int = 10,pbs_length_pooled: int = 14,rtt_max_length_pooled: int = 50,\n",
    "                out_dir: str = './DATETIMESTAMP_PrimeDesign'):\n",
    "    \n",
    "    # Write PrimeDesign Command Line\n",
    "    cmd = 'docker run -v ${PWD}/:/DATA -w /DATA pinellolab/primedesign primedesign_cli' # prefix\n",
    "    cmd += f' -f {file}' # Append required parameters\n",
    "    if pbs_length_list: cmd += f' -pbs {str(np.array(pbs_length_list))[1:-1]}' # Append optional parameters\n",
    "    if rtt_length_list: cmd += f' -rtt {str(np.array(rtt_length_list))[1:-1]}'\n",
    "    if nicking_distance_minimum!=0: cmd += f' -nick_dist_min {str(nicking_distance_minimum)}' \n",
    "    if nicking_distance_maximum!=100: cmd += f' -nick_dist_max {str(nicking_distance_maximum)}'\n",
    "    if filter_c1_extension: cmd += f' -filter_c1 {str(filter_c1_extension)}'\n",
    "    if silent_mutation: cmd += f' - silent_mut'\n",
    "    if genome_wide_design: cmd += f' -genome_wide'\n",
    "    if saturation_mutagenesis: cmd += f' -sat_mut {saturation_mutagenesis}'\n",
    "    if number_of_pegrnas!=3: cmd += f' -n_pegrnas {number_of_pegrnas}'\n",
    "    if number_of_ngrnas!=3: cmd += f' -n_ngrnas {number_of_ngrnas}'\n",
    "    if nicking_distance_pooled!=75: cmd += f' -nick_dist_pooled {nicking_distance_pooled}'\n",
    "    if homology_downstream!=10: cmd += f' -homology_downstream {homology_downstream}'\n",
    "    if pbs_length_pooled!=14: cmd += f' -pbs_pooled {pbs_length_pooled}'\n",
    "    if rtt_max_length_pooled!=50: cmd += f' -rtt_pooled {rtt_max_length_pooled}'\n",
    "    if out_dir!='./DATETIMESTAMP_PrimeDesign': cmd+= f' -out ./DATETIMESTAMP_PrimeDesign'\n",
    "    print(cmd)\n",
    "\n",
    "    os.system(cmd) # Execute PrimeDesign Command Line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'docker run -v ${PWD}/:/DATA -w /DATA pinellolab/primedesign primedesign_cli -f file'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PrimeDesign(file='file')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'docker run -v ${PWD}/:/DATA -w /DATA pinellolab/primedesign primedesign_cli -f file -sat_mut aa'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PrimeDesign(file='file',saturation_mutagenesis='aa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyMUZ.gen.io as io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Target_name</th>\n",
       "      <th>Target_sequence</th>\n",
       "      <th>pegRNA_number</th>\n",
       "      <th>gRNA_type</th>\n",
       "      <th>Spacer_sequence</th>\n",
       "      <th>Spacer_GC_content</th>\n",
       "      <th>PAM_sequence</th>\n",
       "      <th>Extension_sequence</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>...</th>\n",
       "      <th>ngRNA-to-pegRNA_distance</th>\n",
       "      <th>PBS_length</th>\n",
       "      <th>PBS_GC_content</th>\n",
       "      <th>RTT_length</th>\n",
       "      <th>RTT_GC_content</th>\n",
       "      <th>First_extension_nucleotide</th>\n",
       "      <th>Spacer_sequence_order_TOP</th>\n",
       "      <th>Spacer_sequence_order_BOTTOM</th>\n",
       "      <th>pegRNA_extension_sequence_order_TOP</th>\n",
       "      <th>pegRNA_extension_sequence_order_BOTTOM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C2H2-type 2 and 3_1_FtoG</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>pegRNA</td>\n",
       "      <td>CCCGCACTGATTGCACTGGA</td>\n",
       "      <td>0.60</td>\n",
       "      <td>AGG-to-cag</td>\n",
       "      <td>AGAACGGcctggcCAGTGCAATCAGTGC</td>\n",
       "      <td>-</td>\n",
       "      <td>PAM_disrupted_silent_mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.50</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.71</td>\n",
       "      <td>A</td>\n",
       "      <td>caccGCCCGCACTGATTGCACTGGAgtttt</td>\n",
       "      <td>ctctaaaacTCCAGTGCAATCAGTGCGGGC</td>\n",
       "      <td>gtgcAGAACGGcctggcCAGTGCAATCAGTGC</td>\n",
       "      <td>aaaaGCACTGATTGCACTGgccaggCCGTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C2H2-type 2 and 3_1_FtoG</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>tggcCAGTGCAATCAGTGCG</td>\n",
       "      <td>0.60</td>\n",
       "      <td>GGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-nonseed</td>\n",
       "      <td>...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGtggcCAGTGCAATCAGTGCG</td>\n",
       "      <td>aaacCGCACTGATTGCACTGgccaC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C2H2-type 2 and 3_1_FtoG</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>ctggcCAGTGCAATCAGTGC</td>\n",
       "      <td>0.60</td>\n",
       "      <td>GGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-nonseed</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGctggcCAGTGCAATCAGTGC</td>\n",
       "      <td>aaacGCACTGATTGCACTGgccagC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C2H2-type 2 and 3_1_FtoG</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>cctggcCAGTGCAATCAGTG</td>\n",
       "      <td>0.60</td>\n",
       "      <td>CGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-nonseed</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGcctggcCAGTGCAATCAGTG</td>\n",
       "      <td>aaacCACTGATTGCACTGgccaggC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C2H2-type 2 and 3_1_FtoG</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>2</td>\n",
       "      <td>pegRNA</td>\n",
       "      <td>AGGCCCCGCACTGATTGCAC</td>\n",
       "      <td>0.65</td>\n",
       "      <td>TGG-to-tgc</td>\n",
       "      <td>AGAACGGCCCgggCAGTGCAATCAGTGCGGGG</td>\n",
       "      <td>-</td>\n",
       "      <td>PAM_disrupted_silent_mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.64</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>A</td>\n",
       "      <td>caccGAGGCCCCGCACTGATTGCACgtttt</td>\n",
       "      <td>ctctaaaacGTGCAATCAGTGCGGGGCCTC</td>\n",
       "      <td>gtgcAGAACGGCCCgggCAGTGCAATCAGTGCGGGG</td>\n",
       "      <td>aaaaCCCCGCACTGATTGCACTGcccGGGCCGTTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12235</th>\n",
       "      <td>C2H2-type 2 and 3_51_HtoP</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>3059</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>CGTtctCAGGTGGCCAGTGA</td>\n",
       "      <td>0.60</td>\n",
       "      <td>GGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>PE3b-nonseed</td>\n",
       "      <td>...</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGCGTtctCAGGTGGCCAGTGA</td>\n",
       "      <td>aaacTCACTGGCCACCTGagaACGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12236</th>\n",
       "      <td>C2H2-type 2 and 3_51_HtoP</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>3060</td>\n",
       "      <td>pegRNA</td>\n",
       "      <td>CAGCCCCGGACTGCATCCAG</td>\n",
       "      <td>0.70</td>\n",
       "      <td>GGG-to-ggc</td>\n",
       "      <td>CCTGAGGACGcccTCCGGTAGGagcCCTGGATGCAGTCCGGGG</td>\n",
       "      <td>-</td>\n",
       "      <td>PAM_disrupted_silent_mutation</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.71</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.72</td>\n",
       "      <td>C</td>\n",
       "      <td>caccGCAGCCCCGGACTGCATCCAGgtttt</td>\n",
       "      <td>ctctaaaacCTGGATGCAGTCCGGGGCTGC</td>\n",
       "      <td>gtgcCCTGAGGACGcccTCCGGTAGGagcCCTGGATGCAGTCCGGGG</td>\n",
       "      <td>aaaaCCCCGGACTGCATCCAGGgctCCTACCGGAgggCGTCCTCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12237</th>\n",
       "      <td>C2H2-type 2 and 3_51_HtoP</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>3060</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>GGCCACCTGAGGACGcccTC</td>\n",
       "      <td>0.75</td>\n",
       "      <td>CGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-seed</td>\n",
       "      <td>...</td>\n",
       "      <td>-17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGGCCACCTGAGGACGcccTC</td>\n",
       "      <td>aaacGAgggCGTCCTCAGGTGGCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12238</th>\n",
       "      <td>C2H2-type 2 and 3_51_HtoP</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>3060</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>ACCTGAGGACGcccTCCGGT</td>\n",
       "      <td>0.70</td>\n",
       "      <td>AGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-seed</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGACCTGAGGACGcccTCCGGT</td>\n",
       "      <td>aaacACCGGAgggCGTCCTCAGGTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12239</th>\n",
       "      <td>C2H2-type 2 and 3_51_HtoP</td>\n",
       "      <td>AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...</td>\n",
       "      <td>3060</td>\n",
       "      <td>ngRNA</td>\n",
       "      <td>ACGcccTCCGGTAGGagcCC</td>\n",
       "      <td>0.75</td>\n",
       "      <td>TGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>PE3b-seed</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>caccGACGcccTCCGGTAGGagcCC</td>\n",
       "      <td>aaacGGgctCCTACCGGAgggCGTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12240 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Target_name  \\\n",
       "0       C2H2-type 2 and 3_1_FtoG   \n",
       "1       C2H2-type 2 and 3_1_FtoG   \n",
       "2       C2H2-type 2 and 3_1_FtoG   \n",
       "3       C2H2-type 2 and 3_1_FtoG   \n",
       "4       C2H2-type 2 and 3_1_FtoG   \n",
       "...                          ...   \n",
       "12235  C2H2-type 2 and 3_51_HtoP   \n",
       "12236  C2H2-type 2 and 3_51_HtoP   \n",
       "12237  C2H2-type 2 and 3_51_HtoP   \n",
       "12238  C2H2-type 2 and 3_51_HtoP   \n",
       "12239  C2H2-type 2 and 3_51_HtoP   \n",
       "\n",
       "                                         Target_sequence  pegRNA_number  \\\n",
       "0      AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...              1   \n",
       "1      AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...              1   \n",
       "2      AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...              1   \n",
       "3      AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...              1   \n",
       "4      AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...              2   \n",
       "...                                                  ...            ...   \n",
       "12235  AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...           3059   \n",
       "12236  AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...           3060   \n",
       "12237  AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...           3060   \n",
       "12238  AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...           3060   \n",
       "12239  AAACTGCTGATCAGGAAGAGTGGCTCTTCATCACAAAATGTTGCAG...           3060   \n",
       "\n",
       "      gRNA_type       Spacer_sequence  Spacer_GC_content PAM_sequence  \\\n",
       "0        pegRNA  CCCGCACTGATTGCACTGGA               0.60   AGG-to-cag   \n",
       "1         ngRNA  tggcCAGTGCAATCAGTGCG               0.60          GGG   \n",
       "2         ngRNA  ctggcCAGTGCAATCAGTGC               0.60          GGG   \n",
       "3         ngRNA  cctggcCAGTGCAATCAGTG               0.60          CGG   \n",
       "4        pegRNA  AGGCCCCGCACTGATTGCAC               0.65   TGG-to-tgc   \n",
       "...         ...                   ...                ...          ...   \n",
       "12235     ngRNA  CGTtctCAGGTGGCCAGTGA               0.60          GGG   \n",
       "12236    pegRNA  CAGCCCCGGACTGCATCCAG               0.70   GGG-to-ggc   \n",
       "12237     ngRNA  GGCCACCTGAGGACGcccTC               0.75          CGG   \n",
       "12238     ngRNA  ACCTGAGGACGcccTCCGGT               0.70          AGG   \n",
       "12239     ngRNA  ACGcccTCCGGTAGGagcCC               0.75          TGG   \n",
       "\n",
       "                                Extension_sequence Strand  \\\n",
       "0                     AGAACGGcctggcCAGTGCAATCAGTGC      -   \n",
       "1                                              NaN      +   \n",
       "2                                              NaN      +   \n",
       "3                                              NaN      +   \n",
       "4                 AGAACGGCCCgggCAGTGCAATCAGTGCGGGG      -   \n",
       "...                                            ...    ...   \n",
       "12235                                          NaN      -   \n",
       "12236  CCTGAGGACGcccTCCGGTAGGagcCCTGGATGCAGTCCGGGG      -   \n",
       "12237                                          NaN      +   \n",
       "12238                                          NaN      +   \n",
       "12239                                          NaN      +   \n",
       "\n",
       "                          Annotation  ...  ngRNA-to-pegRNA_distance  \\\n",
       "0      PAM_disrupted_silent_mutation  ...                       NaN   \n",
       "1                       PE3b-nonseed  ...                      12.0   \n",
       "2                       PE3b-nonseed  ...                      11.0   \n",
       "3                       PE3b-nonseed  ...                      10.0   \n",
       "4      PAM_disrupted_silent_mutation  ...                       NaN   \n",
       "...                              ...  ...                       ...   \n",
       "12235                   PE3b-nonseed  ...                      -8.0   \n",
       "12236  PAM_disrupted_silent_mutation  ...                       NaN   \n",
       "12237                      PE3b-seed  ...                     -17.0   \n",
       "12238                      PE3b-seed  ...                     -13.0   \n",
       "12239                      PE3b-seed  ...                      -5.0   \n",
       "\n",
       "       PBS_length  PBS_GC_content  RTT_length  RTT_GC_content  \\\n",
       "0            14.0            0.50        14.0            0.71   \n",
       "1             NaN             NaN         NaN             NaN   \n",
       "2             NaN             NaN         NaN             NaN   \n",
       "3             NaN             NaN         NaN             NaN   \n",
       "4            14.0            0.64        18.0            0.72   \n",
       "...           ...             ...         ...             ...   \n",
       "12235         NaN             NaN         NaN             NaN   \n",
       "12236        14.0            0.71        29.0            0.72   \n",
       "12237         NaN             NaN         NaN             NaN   \n",
       "12238         NaN             NaN         NaN             NaN   \n",
       "12239         NaN             NaN         NaN             NaN   \n",
       "\n",
       "       First_extension_nucleotide       Spacer_sequence_order_TOP  \\\n",
       "0                               A  caccGCCCGCACTGATTGCACTGGAgtttt   \n",
       "1                             NaN       caccGtggcCAGTGCAATCAGTGCG   \n",
       "2                             NaN       caccGctggcCAGTGCAATCAGTGC   \n",
       "3                             NaN       caccGcctggcCAGTGCAATCAGTG   \n",
       "4                               A  caccGAGGCCCCGCACTGATTGCACgtttt   \n",
       "...                           ...                             ...   \n",
       "12235                         NaN       caccGCGTtctCAGGTGGCCAGTGA   \n",
       "12236                           C  caccGCAGCCCCGGACTGCATCCAGgtttt   \n",
       "12237                         NaN        caccGGCCACCTGAGGACGcccTC   \n",
       "12238                         NaN       caccGACCTGAGGACGcccTCCGGT   \n",
       "12239                         NaN       caccGACGcccTCCGGTAGGagcCC   \n",
       "\n",
       "         Spacer_sequence_order_BOTTOM  \\\n",
       "0      ctctaaaacTCCAGTGCAATCAGTGCGGGC   \n",
       "1           aaacCGCACTGATTGCACTGgccaC   \n",
       "2           aaacGCACTGATTGCACTGgccagC   \n",
       "3           aaacCACTGATTGCACTGgccaggC   \n",
       "4      ctctaaaacGTGCAATCAGTGCGGGGCCTC   \n",
       "...                               ...   \n",
       "12235       aaacTCACTGGCCACCTGagaACGC   \n",
       "12236  ctctaaaacCTGGATGCAGTCCGGGGCTGC   \n",
       "12237        aaacGAgggCGTCCTCAGGTGGCC   \n",
       "12238       aaacACCGGAgggCGTCCTCAGGTC   \n",
       "12239       aaacGGgctCCTACCGGAgggCGTC   \n",
       "\n",
       "                   pegRNA_extension_sequence_order_TOP  \\\n",
       "0                     gtgcAGAACGGcctggcCAGTGCAATCAGTGC   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                 gtgcAGAACGGCCCgggCAGTGCAATCAGTGCGGGG   \n",
       "...                                                ...   \n",
       "12235                                              NaN   \n",
       "12236  gtgcCCTGAGGACGcccTCCGGTAGGagcCCTGGATGCAGTCCGGGG   \n",
       "12237                                              NaN   \n",
       "12238                                              NaN   \n",
       "12239                                              NaN   \n",
       "\n",
       "                pegRNA_extension_sequence_order_BOTTOM  \n",
       "0                     aaaaGCACTGATTGCACTGgccaggCCGTTCT  \n",
       "1                                                  NaN  \n",
       "2                                                  NaN  \n",
       "3                                                  NaN  \n",
       "4                 aaaaCCCCGCACTGATTGCACTGcccGGGCCGTTCT  \n",
       "...                                                ...  \n",
       "12235                                              NaN  \n",
       "12236  aaaaCCCCGGACTGCATCCAGGgctCCTACCGGAgggCGTCCTCAGG  \n",
       "12237                                              NaN  \n",
       "12238                                              NaN  \n",
       "12239                                              NaN  \n",
       "\n",
       "[12240 rows x 22 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out = io.get('/Users/marczepeda/Documents/Liau_Lab/Projects/2.ZF_degraders/4.Computational/MUZ88/PrimeDesign/240209_17.35.07_PrimeDesign/20240209_05.35.18_PrimeDesign.csv')\n",
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 1, 2]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(np.arange(3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' PrimeDesignOutput: Splits peg/ngRNAs from PrimeDesign output & finishes annotations\n",
    "        primeDesign_output_pt: path to primeDesign output\n",
    "        aa_index: 1st amino acid in target sequence index (Optional, Default: start codon = 1)\n",
    "        scaffold_sequence: sgRNA scaffold sequence (Optional, Default: SpCas9)\n",
    "        epegRNA_motif_sequence: epegRNA motif sequence (Optional, Default: tevopreQ1)\n",
    "    Dependencies: io,numpy\n",
    "'''\n",
    "def PrimeDesignOutput(primeDesign_output_pt: str, aa_index: int=1, \n",
    "                      scaffold_sequence: str='GTTTAAGAGCTATGCTGGAAACAGCATAGCAAGTTTAAATAAGGCTAGTCCGTTATCAACTTGGCTGAATGCCTGCGAGCATCCCACCCAAGTGGCACCGAGTCGGTGC'):\n",
    "    \n",
    "    # Get PrimeDesign output & seperate pegRNAs and ngRNAs\n",
    "    primeDesign_output = io.get(primeDesign_output_pt)\n",
    "    pegRNAs = primeDesign_output[primeDesign_output['gRNA_type']=='pegRNA']\n",
    "    ngRNAs = primeDesign_output[primeDesign_output['gRNA_type']=='ngRNA']\n",
    "\n",
    "    # Generate epegRNAs\n",
    "    pegRNAs['Edit']=[str(target_name.split('_')[2].split('to')[0]) + # AA Before\n",
    "                    str(int(target_name.split('_')[1]) + aa_index-1) + # AA Index\n",
    "                    str(target_name.split('_')[2].split('to')[1]) # AA After\n",
    "                    for target_name in pegRNAs['Target_name']]\n",
    "    pegRNAs['Scaffold_sequence']=[scaffold_sequence]*len(pegRNAs)\n",
    "    pegRNAs['RTT_sequence']=[primeDesign_output.iloc[i]['Extension_sequence'][0:int(primeDesign_output.iloc[i]['RTT_length'])] for i in range(len(pegRNAs))]\n",
    "    pegRNAs['PBS_sequence']=[primeDesign_output.iloc[i]['Extension_sequence'][int(primeDesign_output.iloc[i]['RTT_length']):]  for i in range(len(pegRNAs))]\n",
    "    pegRNAs=pegRNAs[['pegRNA_number','gRNA_type','Strand','Edit', # Important metadata\n",
    "                       'Spacer_sequence','Scaffold_sequence','RTT_sequence','PBS_sequence', # Sequence information\n",
    "                       'Target_name','Target_sequence','Spacer_GC_content','PAM_sequence','Extension_sequence','Annotation','pegRNA-to-edit_distance','Nick_index','ngRNA-to-pegRNA_distance','PBS_length','PBS_GC_content','RTT_length','RTT_GC_content','First_extension_nucleotide']] # Less important metadata\n",
    "\n",
    "    # Generate ngRNAs\n",
    "    ngRNAs['Edit']=[str(target_name.split('_')[2].split('to')[0]) + # AA Before\n",
    "                    str(int(target_name.split('_')[1]) + aa_index-1) + # AA Index\n",
    "                    str(target_name.split('_')[2].split('to')[1]) # AA After\n",
    "                    for target_name in ngRNAs['Target_name']]\n",
    "    ngRNAs['Scaffold_sequence']=[scaffold_sequence]*len(ngRNAs)\n",
    "    ngRNAs['ngRNA_number']=list(np.arange(len(ngRNAs)))\n",
    "    ngRNAs=ngRNAs[['pegRNA_number','ngRNA_number','gRNA_type','Strand','Edit', # Important metadata\n",
    "                   'Spacer_sequence','Scaffold_sequence', # Sequence information\n",
    "                   'Target_name','Target_sequence','Spacer_GC_content','PAM_sequence','Extension_sequence','Annotation','pegRNA-to-edit_distance','Nick_index','ngRNA-to-pegRNA_distance','PBS_length','PBS_GC_content','RTT_length','RTT_GC_content','First_extension_nucleotide']] # Less important metadata\n",
    "\n",
    "    return pegRNAs,ngRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'Levenshtein'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/Users/marczepeda/git/python/other/revise/pe/240831.ipynb Cell 15\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/marczepeda/git/python/other/revise/pe/240831.ipynb#X23sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpegLIT\u001b[39;00m\n",
      "File \u001b[0;32m~/git/python/other/revise/pe/pegLIT.py:7\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mspecial\u001b[39;00m \u001b[39mimport\u001b[39;00m expit \u001b[39mas\u001b[39;00m sigmoid\n\u001b[1;32m      6\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39msklearn\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcluster\u001b[39;00m \u001b[39mimport\u001b[39;00m AgglomerativeClustering \u001b[39mas\u001b[39;00m HAC\n\u001b[0;32m----> 7\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mLevenshtein\u001b[39;00m \u001b[39mimport\u001b[39;00m distance \u001b[39mas\u001b[39;00m levenshtein_distance\n\u001b[1;32m      8\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mRNA\u001b[39;00m \u001b[39m# ViennaRNA\u001b[39;00m\n\u001b[1;32m     10\u001b[0m BASE_SYMBOLS \u001b[39m=\u001b[39m {\n\u001b[1;32m     11\u001b[0m \u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m,), \u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m,), \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m,), \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m,), \u001b[39m\"\u001b[39m\u001b[39mU\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m,),\n\u001b[1;32m     12\u001b[0m \u001b[39m\"\u001b[39m\u001b[39mW\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mS\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mM\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m),\n\u001b[1;32m     13\u001b[0m \u001b[39m\"\u001b[39m\u001b[39mK\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mR\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mY\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m),\n\u001b[1;32m     14\u001b[0m \u001b[39m\"\u001b[39m\u001b[39mB\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mD\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mH\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m), \u001b[39m\"\u001b[39m\u001b[39mV\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m),\n\u001b[1;32m     15\u001b[0m \u001b[39m\"\u001b[39m\u001b[39mN\u001b[39m\u001b[39m\"\u001b[39m: (\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mC\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m)}\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'Levenshtein'"
     ]
    }
   ],
   "source": [
    "import pegLIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' epegRNA_linkers: Generate epegRNA linkers between PBS and motif\n",
    "        pegRNAs:\n",
    "        epegRNA_motif_sequence:\n",
    "        checkpoint_out: optional\n",
    "        checkpoint_file: optional\n",
    "        Need to make checkpoints optional for the epegRNAs\n",
    "    Dependencies: pandas,pegLIT\n",
    "'''\n",
    "def epegRNA_linkers(pegRNAs: pd.DataFrame, epegRNA_motif_sequence: str='CGCGGTTCTATCTAGTTACGCGTTAAACCAACTAGAA',\n",
    "                    out: str=None, checkpoint_file=None):\n",
    "    \n",
    "    linkers = []\n",
    "    checkpoint = pd.DataFrame(columns=['id','linker'])\n",
    "    for i in range(len(pegRNAs)):\n",
    "        linkers.extend(pegLIT.pegLIT(seq_spacer=epegRNAs_input.iloc[i]['spacer'],\n",
    "                                seq_scaffold=epegRNAs_input.iloc[i]['scaffold'],\n",
    "                                seq_template=epegRNAs_input.iloc[i]['RTT'],\n",
    "                                seq_pbs=epegRNAs_input.iloc[i]['PBS'],\n",
    "                                seq_motif=epegRNAs_input.iloc[i]['motif']))\n",
    "        \n",
    "        # Save checkpoints\n",
    "        temp = pd.DataFrame({'id': [epegRNAs_input.iloc[i]['id']],\n",
    "                            'linker': [linkers[i]]})\n",
    "        checkpoint = pd.concat([checkpoint,temp])\n",
    "        csv.save_csv(dir='/Users/marczepeda/Documents/Liau_Lab/Projects/2.ZF_degraders/4.Computational/MUZ88/240212_guide_analysis_all',\n",
    "                    file='epegRNAs_linkers_checkpoint.csv',\n",
    "                    df=checkpoint,\n",
    "                    cols=checkpoint.columns)\n",
    "        \n",
    "    epegRNAs_input['linker'] = linkers\n",
    "    epegRNAs_input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "''' PrimeDesignFeatures: Distill PE library into spacer and PBS motifs.\n",
    "        ...\n",
    "    Dependencies:\n",
    "'''\n",
    "def PrimeDesignFeatures():\n",
    "    return"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
